### PR TITLE
Or 1927 update listener withdrawal flow for thanos titan app listen on optimism portal2 and l2 to l1 message passer

### PR DESCRIPTION
**Pull request by nguyenzung**
Handle withdrawal event in primitive contracts: OptimismPortal and L2ToL1MessagePasser

Thank you!

### CODE REVIEW

**Comment by nguyenzung** [comment]
I think we should put notifier in the callback function. It is great if handing notifier by the callback, because we don't need to push message to Slack everytime an event occurs. 

https://github.com/tokamak-network/tokamak-thanos-event-listener/blob/OR-1927-update-listener-withdrawal-flow-for-thanos-titan-app-listen-on-OptimismPortal2-and-L2ToL1MessagePasser/internal/pkg/listener/request.go#L49

File: internal/app/thanos-notif/withdrawal.go [hunk 1]
```diff
@@ -10,6 +10,60 @@ import (
...
+	if err != nil {
+		return "", "", err
+	}
+	event, err := l2ToL1MessagePasserFilterer.ParseMessagePassed(*vLog)
+	if err != nil {
+		log.GetLogger().Errorw("MessagePassed event parsing fail", "error", err)
+		return "", "", err
+	}
+	// Slack notify title and text
+	title := fmt.Sprintf("[" + p.cfg.Network + "] [Withdrawal Finalized]")
```
Comment (ghost) [inline internal/app/thanos-notif/withdrawal.go]: This line must be "Withdrawal Initiated", right?

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Please consider this case: we withdraw by the standard bridge, and we are listening to two events:
- `MessagePassed` from the `L2ToL1MessagePasser` contract
- `WithdrawalInitiated` from the `L2StandardBridge` contract

-> So, we can duplicate the events from the same tx.

File: internal/app/thanos-notif/withdrawal.go (hunk 1 above)
Comment (nguyenzung) [reply internal/app/thanos-notif/withdrawal.go]: Totally correct. Thank you!

**Comment by nguyenzung** [comment]
> Please consider this case: we withdraw by the standard bridge, and we are listening to two events:
> 
> * `MessagePassed` from the `L2ToL1MessagePasser` contract
> * `WithdrawalInitiated` from the `L2StandardBridge` contract
> 
> -> So, we can duplicate the events from the same tx.

These event actually are not the same. MessagePassed is used for all withdrawals. WithdrawalInitiated is used with only withdrawal that triggered by L2StandardBridge, including some other event info.

**Comment by ghost** [comment]
> > Please consider this case: we withdraw by the standard bridge, and we are listening to two events:
> > 
> > * `MessagePassed` from the `L2ToL1MessagePasser` contract
> > * `WithdrawalInitiated` from the `L2StandardBridge` contract
> > 
> > -> So, we can duplicate the events from the same tx.
> 
> These event actually are not the same. MessagePassed is used for all withdrawals. WithdrawalInitiated is used with only withdrawal that triggered by L2StandardBridge, including some other event info.

Yes, they aren't the same but we will trigger two notifications every time `withdraw initiated` when bridging through the L2StandardBridge contract.

**Comment by nguyenzung** [comment]
> > > Please consider this case: we withdraw by the standard bridge, and we are listening to two events:
> > > 
> > > * `MessagePassed` from the `L2ToL1MessagePasser` contract
> > > * `WithdrawalInitiated` from the `L2StandardBridge` contract
> > > 
> > > -> So, we can duplicate the events from the same tx.
> > 
> > 
> > These event actually are not the same. MessagePassed is used for all withdrawals. WithdrawalInitiated is used with only withdrawal that triggered by L2StandardBridge, including some other event info.
> 
> Yes, they aren't the same but we will trigger two notifications every time `withdraw initiated` when bridging through the L2StandardBridge contract.

I understood your opinion. I mean we can keep these events. Because StandardBridges have important events needed to notify, and the same thing for our Portals. 

What do you think about it? @steven94kr @theo-learner @ohbyeongmin @xxeonge

**Comment by steven94kr** [comment]
I think we can listen for Event but there should be only one notification.

It can be very confusing to receive two notifications for your deposit.

I think it would be best if we could find a way. However, if it is difficult to parse MassagePasser events, it may be a good idea to use the current version.

**Comment by nguyenzung** [comment]
> I think we can listen for Event but there should be only one notification.
> 
> It can be very confusing to receive two notifications for your deposit.
> 
> I think it would be best if we could find a way. However, if it is difficult to parse MassagePasser events, it may be a good idea to use the current version.

I think it could be a good approach. Let me think about it. Thank you!

**Comment by nguyenzung** [comment]
Hi @steven94kr @0x6e616d @theo-learner 
Please check my updates

**Review by theo-learner — APPROVED** [review]
Thank you so much, Brave!

**Pull request merged**